sys/malloc.c

```diff
@@ -184,12 +194,20 @@ void *kmalloc(malloc_pool_t *mp, size_t size, uint16_t flags) {
     if (mb) {
       if (flags == M_ZERO)
         memset(mb->mb_data, 0, size);
-
+      mtx_unlock(&mp->mutex);
       return mb->mb_data;
     }
   }
-
   /* Couldn't find any continuous memory with the requested size. */
+  if (mp->pages_used < mp->pages_limit) {
```

Comment:
If flag `M_NOWAIT` is passed to `kmalloc` you should not attempt to allocate new pages. In future `kmalloc_add_pages` may trigger page replacement algorithm, that may force a thread go sleep!